runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingModeExecutionContext.java

```diff
@@ -581,30 +594,96 @@ public void flushState() {
+      }
+    }
+
+    // A list of timer keys that were modified by user processing earlier in this bundle. This
+    // serves a tombstone, so
+    // that we know not to fire any bundle tiemrs that were moddified.
+    private Table<String, StateNamespace, TimerData> modifiedUserTimerKeys = null;
+
+    private void onUserTimerModified(TimerData timerData) {
+      if (!timerData.getDeleted()) {
+        getModifiedUserTimersOrdered(timerData.getDomain()).add(timerData);
+      }
+      modifiedUserTimerKeys.put(
+          WindmillTimerInternals.getTimerDataKey(timerData), timerData.getNamespace(), timerData);
+    }
+
+    private boolean timerModified(TimerData timerData) {
+      String timerKey = WindmillTimerInternals.getTimerDataKey(timerData);
+      @Nullable
+      TimerData updatedTimer = modifiedUserTimerKeys.get(timerKey, timerData.getNamespace());
+      return updatedTimer != null && !updatedTimer.equals(timerData);
+    }
 
     public <W extends BoundedWindow> TimerData getNextFiredUserTimer(Coder<W> windowCoder) {
       if (cachedFiredUserTimers == null) {
+        // This is the first call to getNextFiredUserTimer in this bundle. Extract any user timers
+        // from the bundle
+        // and cache the list for the rest of this bundle processing.
         cachedFiredUserTimers =
-            FluentIterable.<Timer>from(StreamingModeExecutionContext.this.getFiredTimers())
-                .filter(
-                    timer ->
-                        WindmillTimerInternals.isUserTimer(timer)
-                            && timer.getStateFamily().equals(stateFamily))
-                .transform(
-                    timer ->
-                        WindmillTimerInternals.windmillTimerToTimerData(
-                            WindmillNamespacePrefix.USER_NAMESPACE_PREFIX, timer, windowCoder))
-                .iterator();
+            Iterators.peekingIterator(
+                FluentIterable.from(StreamingModeExecutionContext.this.getFiredTimers())
+                    .filter(
+                        timer ->
+                            WindmillTimerInternals.isUserTimer(timer)
+                                && timer.getStateFamily().equals(stateFamily))
+                    .transform(
+                        timer ->
+                            WindmillTimerInternals.windmillTimerToTimerData(
+                                WindmillNamespacePrefix.USER_NAMESPACE_PREFIX, timer, windowCoder))
+                    .iterator());
       }
 
-      if (!cachedFiredUserTimers.hasNext()) {
-        return null;
+      while (cachedFiredUserTimers.hasNext()) {
+        TimerData nextInBundle = cachedFiredUserTimers.peek();
+        NavigableSet<TimerData> modifiedUserTimersOrdered =
+            getModifiedUserTimersOrdered(nextInBundle.getDomain());
+        // If there is a modified timer that is earlier than the next timer in the bundle, try and
```

Comment:
Reading kind of quick so I probably missed something, but can `cachedFiredUserTimers` just be a navigable set and when you set a `modifiedUserTimer` you insert into it? (hence the earliest eligible timer is always the one to fire next)

Is this a cost thing or semantics?

Reply:
So that was my initial implementation, and it does work. I changed it to this structure for two main reasons
   1. This code structure mirrors the one in FnApiDoFnRunner (where timers are presented one at a time) and I preferred using the same algorithm in both places.
   2.  The other approach added a bit more code complexity (e.g. deciding when a new timer should be inserted into cachedFiredUser Timers - timers far in the future should not be). I found this code to be simpler to follow.